<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="Aurelien Ginolhac, Eric Koncina" />


<title>practical broom</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<link href="site_libs/tutorial-0.1/css/style.css" rel="stylesheet" />
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-1.1/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-1.1/highlight.js"></script>
<link href="site_libs/font-awesome-4.5.0/css/font-awesome.min.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs && document.readyState && document.readyState === "complete") {
   window.setTimeout(function() {
      hljs.initHighlighting();
   }, 0);
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>

<link rel="stylesheet" href="css/tutorial.css" type="text/css" />

</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}

.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->






<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">R tidyverse</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-graduation-cap"></span>
     
    Lectures
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="lecture01_intro.html">Lecture 01 - intro</a>
    </li>
    <li>
      <a href="lecture02_data_structure.html">Lecture 02 - data types &amp; structures</a>
    </li>
    <li>
      <a href="lecture03_rmarkdown.html">Lecture 03 - Rmarkdown</a>
    </li>
    <li>
      <a href="lecture04_readr.html">Lecture 04 - importing data</a>
    </li>
    <li>
      <a href="lecture05_tidyr.html">Lecture 05 - tidy data</a>
    </li>
    <li>
      <a href="lecture06_dplyr.html">Lecture 06 - manipulation with dplyr</a>
    </li>
    <li>
      <a href="lecture07_plotting.html">Lecture 07 - data plotting</a>
    </li>
    <li>
      <a href="lecture08_purrr.html">Lecture 08 - functional programming with purrr</a>
    </li>
    <li>
      <a href="lecture09_broom.html">Lecture 09 - data handling with broom</a>
    </li>
    <li>
      <a href="lecture10_tidyeval.html">Lecture 10 - non standard evaluation with tidyeval</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">slides, keyboard shortcuts:</li>
    <li class="dropdown-header">o: outline</li>
    <li class="dropdown-header">w: widescreen on/off</li>
    <li class="dropdown-header">f: fullscreen on/off</li>
    <li class="dropdown-header">ctrl +/ ctrl -: zoom in/out</li>
  </ul>
</li>
<li>
  <a href="practicals.html">
    <span class="fa fa-desktop"></span>
     
    Practicals
  </a>
</li>
<li>
  <a href="install_tutorial.html">
    <span class="fa fa-cogs"></span>
     
    Install
  </a>
</li>
<li>
  <a href="misc.html">
    <span class="fa fa-book"></span>
     
    Misc
  </a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="about.html">
    <span class="fa fa-question fa-lg"></span>
     
    About
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">practical broom</h1>
<h3 class="subtitle"><em>Yeast dataset</em></h3>
<h4 class="author"><em>Aurelien Ginolhac, Eric Koncina</em></h4>
<h4 class="date"><em>4 May 2017</em></h4>

</div>


<blockquote>
<p>This practical work is adapted from the exhaustive example published by David Robinson on <a href="http://varianceexplained.org/r/tidy-genomics/">his blog</a>.</p>
</blockquote>
<p>In <a href="http://www.molbiolcell.org/content/19/1/352.abstract">2008, Brauer et al.</a> used microarrays to test the effect of starvation on the growth rate of yeast. For example, they tried limiting the yeast’s supply of glucose (sugar to metabolize into energy), of leucine (an essential amino acid), or of ammonium (a source of nitrogen).</p>
<div id="project---set-up" class="section level2">
<h2>Project - set-up</h2>
<ul>
<li>Create a folder <code>data</code></li>
<li>Download <a href="https://lsru.github.io/r_workshop/data/Brauer2008_DataSet1.tds"><code>Brauer2008_DataSet1.tds</code></a> inside the <code>data</code> folder</li>
</ul>
<p>Load the [<code>Brauer2008_DataSet1.tds</code>] file as a <code>tibble</code> named <code>original_data</code>. This is the exact data that was published with the paper (though for some reason the link on the journal’s page is broken). It thus serves as a good example of tidying a biological dataset “found in the wild”.</p>
</div>
<div id="tidying-the-data" class="section level1">
<h1>1 Tidying the data</h1>
<p>Have a look at the dataset. Is the data “tidy”?</p>
<div id="many-variables-are-stored-in-one-column" class="section level2">
<h2>1.1 Many variables are stored in one column</h2>
<ul>
<li><strong>Gene name</strong> e.g. <em>SFB2</em>. Note that not all genes have a name.</li>
<li><strong>Biological process</strong> e.g. “proteolysis and peptidolysis”</li>
<li><strong>Molecular function</strong> e.g. “metalloendopeptidase activity”</li>
<li><strong>Systematic ID</strong> e.g. <em>YNL049C</em>. Unlike a gene name, every gene in this dataset has a systematic ID.</li>
<li><strong>Another ID number</strong> e.g. <code>1082129</code>. We don’t know what this number means, and it’s not annotated in the paper. Oh, well.</li>
</ul>
<ol style="list-style-type: decimal">
<li>Use the appropriate function provided in the <code>tidyr</code> library to split these values and generate a column for each variable.</li>
</ol>
<div class="box" style="background-color:rgba(173, 216, 230, 0.3);border:2px solid rgba(173, 216, 230, 1);">
<h2 style="background-color:rgba(173, 216, 230, 1);">
Tip
</h2>
<div class="box-content">
<p>Special characters such as pipes (<code>|</code>) must be despecialized as they have a specific meaning. In R, you have to use 2 backslashes like <code>\\|</code> for one pipe <code>|</code></p>
</div>
</div>
<ol start="2" style="list-style-type: decimal">
<li>Once you separated the variables delimited by two “<code>||</code>”, check closer the new values: You will see that they might start and/or end with whitespaces which might be inconvinient during the subsequent use.
<ul>
<li>To remove these whitespaces, R base provides a function called <code>trimws()</code>. Let’s test how the function works:</li>
<li><code>dplyr</code> allows us to apply a function (in our case <code>trimws()</code>) to all columns. In other words, we would like to modify the content of each column with the output of the function <code>trimws()</code>. How can you achieve this? Save the result in a <code>tibble</code> called <code>cleaned_data</code>.</li>
</ul></li>
</ol>
<!-- Moving chunk down as it breaks the indentation in HTML (Is it working in Pdf?) -->
<pre class="r"><code># Creating test string with whitespaces:
s &lt;- &quot;  Removing whitespaces at both ends &quot;
s</code></pre>
<pre><code>## [1] &quot;  Removing whitespaces at both ends &quot;</code></pre>
<pre class="r"><code>trimws(s)</code></pre>
<pre><code>## [1] &quot;Removing whitespaces at both ends&quot;</code></pre>
<ol start="3" style="list-style-type: decimal">
<li>We are not going to use every column of the dataframe. Remove the unnecessary columns: <code>number</code>, <code>GID</code>, <code>YORF</code> and <code>GWEIGHT</code>.</li>
</ol>
<p>Look at the column names.<br />
Do you think that our dataset is now “tidy”?</p>
</div>
<div id="column-headers-are-values-not-variable-names" class="section level2">
<h2>1.2 Column headers are values, not variable names</h2>
<ul>
<li>Keep care to build a tibble with each column representing a variable: At this point we are storing the sample name as a different column <code>sample</code> associated to values in <code>expression</code> column. Save as <code>cleaned_data_melt</code></li>
</ul>
<p>Now look at the content of the <code>sample</code> column. We are again facing the problem that two variables are stored in a single column.</p>
<p>Use the same function as before to split the <code>sample</code> column into two variables <code>nutrient</code> and <code>rate</code> (use the appropriate delimitation in <code>sep</code>.</p>
<div class="box" style="background-color:rgba(173, 216, 230, 0.3);border:2px solid rgba(173, 216, 230, 1);">
<h2 style="background-color:rgba(173, 216, 230, 1);">
Tip
</h2>
<div class="box-content">
<p>Consider using the <code>convert</code> argument. It allows to convert strings to number when relevant like here.</p>
</div>
</div>
</div>
<div id="turn-nutrient-letters-into-more-comprehensive-words" class="section level2">
<h2>1.3 Turn nutrient letters into more comprehensive words</h2>
<p>Right now, the nutrients are designed by a single letter. It would be nice to have the full word instead. One could use a full mixture of <code>if</code> and <code>else</code> such as <code>if_else(nutrient == &quot;G&quot;, &quot;Glucose&quot;, if_else(nutrient == &quot;L&quot;, &quot;Leucine&quot;, etc ...))</code> But, that would be cumbersome.</p>
<p>using the following correspondences and <code>dplyr::recode</code>, recode all nutrient names.</p>
<pre class="r"><code>G = &quot;Glucose&quot;, L = &quot;Leucine&quot;, P = &quot;Phosphate&quot;,
S = &quot;Sulfate&quot;, N = &quot;Ammonia&quot;, U = &quot;Uracil&quot;</code></pre>
</div>
<div id="cleaning-up-missing-data" class="section level2">
<h2>1.4 Cleaning up missing data</h2>
<p>Two variables must be present for the further analysis:</p>
<ul>
<li>gene expression</li>
<li>systematic_name</li>
</ul>
<p>delete observations that are missing any of the two mandatory variables. How many rows did you remove?</p>
</div>
</div>
<div id="representing-the-data" class="section level1">
<h1>2 Representing the data</h1>
<p>Tidying the data is a crucial step allowing easy handling and representing.</p>
<div id="plot-the-expression-data-of-the-leu1-gene" class="section level2">
<h2>2.1 Plot the expression data of the <em>LEU1</em> gene</h2>
<p>Extract the data corresponding to the gene called <em>LEU1</em> and draw a line for each nutrient showing the expression in function of the growth rate.</p>
</div>
<div id="plot-the-expression-data-of-a-biological-process" class="section level2">
<h2>2.2 Plot the expression data of a biological process</h2>
<p>For this, we don’t need to filter by single gene names as the raw data provides us some information on the biological process for each gene.<br />
Extract all the genes in the <strong>leucine biosynthesis</strong> process and plot the expression in function of the growth rate for each nutrient.</p>
</div>
<div id="perform-a-linear-regression-in-top-of-the-plots" class="section level2">
<h2>2.3 Perform a linear regression in top of the plots</h2>
<p>Let’s play with the graph a little more. These trends look vaguely linear.<br />
Add a linear regression with the appropriate <code>ggplot2</code> function and carrefully adjust the <code>method</code> argument.</p>
</div>
<div id="switch-to-another-biological-process" class="section level2">
<h2>2.4 Switch to another biological process</h2>
<p>Once the dataset is tidy, it is very easy to switch to another biological process. Instead of the “leucine biosynthesis”, plot the data corresponding to <strong>sulfur metabolism</strong>.</p>
<div class="box" style="background-color:rgba(173, 216, 230, 0.3);border:2px solid rgba(173, 216, 230, 1);">
<h2 style="background-color:rgba(173, 216, 230, 1);">
Tip
</h2>
<div class="box-content">
<p>you can combine the facet headers using <code>+</code> in <code>facet_wrap()</code>. Adding the systematic name allows to get a name when the gene name is missing.</p>
</div>
</div>
</div>
</div>
<div id="linear-models" class="section level1">
<h1>3. Linear models</h1>
<p>We can see that most genes follow a linear trend under a starvation stress. Instead of manipulating all data points, we would like to estimate just the 2 parameters of a linear model. What are they? And what do they represent for this experiment?</p>
<div id="perform-all-linear-models" class="section level2">
<h2>3.1 Perform all linear models</h2>
<p>Perform a linear regression of <code>expression</code> explained by <code>rate</code> for all group of name, systematic_name and nutrient. By sure to convert the linear models to a <code>tibble</code> using <code>broom</code> as saved as <code>cleaned_lm</code></p>
<p>The computation takes ~ 60 sec on my macbook pro. For test purposes, you may want to subsample per group a small number of genes</p>
<ul>
<li><p>How many models did you perform?</p></li>
<li><p>bonus question: you have observed a warning <code>essentially perfect fit: summary may be unreliable</code>, why? How can we clean up the dataset to avoid it?</p></li>
</ul>
</div>
<div id="explore-models-intercept-values" class="section level2">
<h2>3.2 Explore models, intercept values</h2>
<p>For the gene <em>LEU1</em>, let’s look again at the linear models per nutrient and add as a dashed black line for mean of all intercepts.</p>
<div class="figure">
<img src="TD_broom_files/figure-html/leu1_intercept-1.png" alt="LEU1 linear models per nutrient" width="672" />
<p class="caption">
LEU1 linear models per nutrient
</p>
</div>
<p>We can see that when <strong>leucine</strong> is the limiting factor, the yeast expresses massively this gene <em>LEU1</em> to compensate. Remember that all experiments are performed at constant growth rates in this chemostat.</p>
<p>Now, compute the difference between the intercept and the mean(intercept) per systematic name, so for all nutrient per gene. And select the top 20 highest intercept values to their means. Save as <code>top_20_intercept</code></p>
<ul>
<li>merge those 20 genes with the <code>cleaned_data_melt</code> to retreived the original data points</li>
<li><p>plot the data points and linear trends for those top 20 genes. Instead of the systematic name in the header, use the molecular function (<code>MF</code>)</p></li>
<li><p>what can conclude?</p></li>
</ul>
</div>
<div id="explore-models-slope-estimates" class="section level2">
<h2>3.3 Explore models, slope estimates</h2>
<p>We can now have a look at the slope estimates.</p>
<ul>
<li><p>first, filter the term for slope estimate <code>rate</code> and remove <em>p-value</em> that are missing. Save as <code>rate_slopes</code></p></li>
<li><p>display the <em>p-values</em> histograms per nutrient</p></li>
<li><p>looking at those histograms, how do you estimate the ratio of <em>null</em> and <em>alternative</em> hypotheses?</p></li>
<li><p>how can retrieved the genes that belong most probably to the _alternative hypothesis?</p></li>
<li><p>display the histograms of both <em>p</em> and <em>q</em> <em>values</em> per nutrient</p></li>
</ul>
</div>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
