<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="Roland Krause" />


<title>Practical data transformation</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<link href="site_libs/tutorial-0.1/css/style.css" rel="stylesheet" />
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-1.1/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-1.1/highlight.js"></script>
<link href="site_libs/font-awesome-4.5.0/css/font-awesome.min.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs && document.readyState && document.readyState === "complete") {
   window.setTimeout(function() {
      hljs.initHighlighting();
   }, 0);
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>

<link rel="stylesheet" href="css/tutorial.css" type="text/css" />

</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}

.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->






<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">R tidyverse</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-graduation-cap"></span>
     
    Lectures
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="lecture01_intro.html">Lecture 01 - intro</a>
    </li>
    <li>
      <a href="lecture02_data_structure.html">Lecture 02 - data types &amp; structures</a>
    </li>
    <li>
      <a href="lecture03_rmarkdown.html">Lecture 03 - Rmarkdown</a>
    </li>
    <li>
      <a href="lecture04_readr.html">Lecture 04 - importing data</a>
    </li>
    <li>
      <a href="lecture05_tidyr.html">Lecture 05 - tidy data</a>
    </li>
    <li>
      <a href="lecture06_dplyr.html">Lecture 06 - manipulation with dplyr</a>
    </li>
    <li>
      <a href="lecture07_plotting.html">Lecture 07 - data plotting</a>
    </li>
    <li>
      <a href="lecture08_purrr.html">Lecture 08 - functional programming with purrr</a>
    </li>
    <li>
      <a href="lecture09_broom.html">Lecture 09 - data handling with broom</a>
    </li>
    <li>
      <a href="lecture10_tidyeval.html">Lecture 10 - non standard evaluation with tidyeval</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">slides, keyboard shortcuts:</li>
    <li class="dropdown-header">o: outline</li>
    <li class="dropdown-header">w: widescreen on/off</li>
    <li class="dropdown-header">f: fullscreen on/off</li>
    <li class="dropdown-header">ctrl +/ ctrl -: zoom in/out</li>
  </ul>
</li>
<li>
  <a href="practicals.html">
    <span class="fa fa-desktop"></span>
     
    Practicals
  </a>
</li>
<li>
  <a href="install_tutorial.html">
    <span class="fa fa-cogs"></span>
     
    Install
  </a>
</li>
<li>
  <a href="misc.html">
    <span class="fa fa-book"></span>
     
    Misc
  </a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="about.html">
    <span class="fa fa-question fa-lg"></span>
     
    About
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Practical data transformation</h1>
<h3 class="subtitle"><em>dplyr</em></h3>
<h4 class="author"><em>Roland Krause</em></h4>
<h4 class="date"><em>3 May 2017</em></h4>

</div>


<p>This short tutorial will allow you to explore <code>dplyr</code> functionality based on the previous lecture. Every question can be answered with a combination of <code>%&gt;%</code> pipes. You should refrain from using temporary variables and statements outside of the range of the tidyverse.</p>
<div class="box" style="background-color:rgba(173, 216, 230, 0.3);border:2px solid rgba(173, 216, 230, 1);">
<h2 style="background-color:rgba(173, 216, 230, 1);">
Tip
</h2>
<div class="box-content">
<p>If you are missing `biomaRt, install with</p>
<pre><code>source(&quot;https://bioconductor.org/biocLite.R&quot;)
biocLite(&quot;biomaRt&quot;)</code></pre>
</div>
</div>
<ol class="example" style="list-style-type: decimal">
<li>Get data for chromosome 21 from biomaRt. Use Rmarkdown and set the chunk option ‘cache = TRUE’</li>
</ol>
<pre class="r"><code>library(biomaRt)</code></pre>
<pre><code>## Loading required package: methods</code></pre>
<pre class="r"><code>gene_mart &lt;- useMart(biomart=&quot;ENSEMBL_MART_ENSEMBL&quot;, host=&quot;www.ensembl.org&quot;)
gene_set &lt;- useDataset(gene_mart , dataset=&quot;hsapiens_gene_ensembl&quot;)

gene_by_exon &lt;-as_tibble(getBM(
  mart = gene_set,
  attributes = c(
    &quot;ensembl_gene_id&quot;,
    &quot;ensembl_transcript_id&quot;,
    &quot;ensembl_exon_id&quot;,
    &quot;chromosome_name&quot;,
    &quot;start_position&quot;,
    &quot;end_position&quot;,
    &quot;hgnc_symbol&quot;,
    &quot;hgnc_id&quot;,
    &quot;strand&quot;,
    &quot;gene_biotype&quot;,
    &quot;phenotype_description&quot;
    ), 
  filter = &quot;chromosome_name&quot;,
  value = &quot;21&quot;
  ))</code></pre>
<ol start="2" class="example" style="list-style-type: decimal">
<li>Extract the processed pseudogenes from the <code>genes_by_exon</code> data set. Convert <code>genes_by_exon</code> data set to a <code>tibble</code> Use <code>glimpse()</code> to find the correct columns and <code>distinct()</code> to identify how pseudogenes are coded. Store the results in a tibble <code>pseudogenes</code></li>
</ol>
<div class="box solution">
<h2>
Solution
</h2>
<div class="box-content">
<pre class="r"><code>gene_by_exon &lt;- gene_by_exon %&gt;% 
  as_tibble()

pseudogenes &lt;- gene_by_exon %&gt;% 
  filter(gene_biotype == &quot;processed_pseudogene&quot;)</code></pre>
</div>
</div>
<ol start="3" class="example" style="list-style-type: decimal">
<li>Count the number of pseudogenes in the set (without referring to <code>table()</code> obviously )</li>
</ol>
<div class="box solution">
<h2>
Solution
</h2>
<div class="box-content">
<pre class="r"><code>pseudogenes %&gt;% count()</code></pre>
<pre><code>## # A tibble: 1 × 1
##       n
##   &lt;int&gt;
## 1   170</code></pre>
</div>
</div>
<ol start="4" class="example" style="list-style-type: decimal">
<li>Extract a unique set of gene ids without redunancy of transcripts and exon information. Store the results in a tibble called <code>genes</code></li>
</ol>
<div class="box solution">
<h2>
Solution
</h2>
<div class="box-content">
<pre class="r"><code>gene_by_exon %&gt;% 
  dplyr::select(-ensembl_transcript_id, -ensembl_exon_id) %&gt;% 
  distinct()-&gt; genes</code></pre>
</div>
</div>
<ol start="5" class="example" style="list-style-type: decimal">
<li>Sort the genes by their length.</li>
</ol>
<div class="box solution">
<h2>
Solution
</h2>
<div class="box-content">
<pre class="r"><code>gene_by_exon %&gt;%
  mutate(length=end_position-start_position) %&gt;% 
  arrange(length)</code></pre>
<pre><code>## # A tibble: 19,330 × 12
##    ensembl_gene_id ensembl_transcript_id ensembl_exon_id chromosome_name
##              &lt;chr&gt;                 &lt;chr&gt;           &lt;chr&gt;           &lt;int&gt;
## 1  ENSG00000266692       ENST00000581669 ENSE00002724833              21
## 2  ENSG00000275523       ENST00000611656 ENSE00003745719              21
## 3  ENSG00000264063       ENST00000577708 ENSE00002724720              21
## 4  ENSG00000277437       ENST00000614492 ENSE00003754026              21
## 5  ENSG00000275167       ENST00000611994 ENSE00003717901              21
## 6  ENSG00000283904       ENST00000385060 ENSE00001500066              21
## 7  ENSG00000284448       ENST00000290239 ENSE00003729123              21
## 8  ENSG00000275166       ENST00000622292 ENSE00003738277              21
## 9  ENSG00000201025       ENST00000364155 ENSE00001438918              21
## 10 ENSG00000263681       ENST00000582241 ENSE00002698395              21
## # ... with 19,320 more rows, and 8 more variables: start_position &lt;int&gt;,
## #   end_position &lt;int&gt;, hgnc_symbol &lt;chr&gt;, hgnc_id &lt;chr&gt;, strand &lt;int&gt;,
## #   gene_biotype &lt;chr&gt;, phenotype_description &lt;chr&gt;, length &lt;int&gt;</code></pre>
</div>
</div>
<ol start="6" class="example" style="list-style-type: decimal">
<li>Calculate the average length per gene by <code>gene_biotype</code>.</li>
</ol>
<div class="box solution">
<h2>
Solution
</h2>
<div class="box-content">
<pre class="r"><code>gene_by_exon %&gt;%
  mutate(length=end_position-start_position) %&gt;% 
  group_by(gene_biotype) %&gt;% 
  summarize(mean_lenght = floor(mean(length)))</code></pre>
<pre><code>## # A tibble: 19 × 2
##                          gene_biotype mean_lenght
##                                 &lt;chr&gt;       &lt;dbl&gt;
## 1            3prime_overlapping_ncRNA        3643
## 2                           antisense       34382
## 3                           IG_V_gene         435
## 4                             lincRNA      105951
## 5                               miRNA          86
## 6                            misc_RNA         145
## 7                processed_pseudogene        1006
## 8                processed_transcript       72905
## 9                      protein_coding      139095
## 10                         pseudogene         103
## 11                               rRNA         132
## 12                     sense_intronic       12055
## 13                  sense_overlapping       40989
## 14                             snoRNA         118
## 15                              snRNA         116
## 16                                TEC        2413
## 17   transcribed_processed_pseudogene       97124
## 18 transcribed_unprocessed_pseudogene       84268
## 19             unprocessed_pseudogene       12495</code></pre>
</div>
</div>
<ol start="7" class="example" style="list-style-type: decimal">
<li>Calculate the total number of genes and their average length by <code>gene_biotype</code>.</li>
</ol>
<div class="box solution">
<h2>
Solution
</h2>
<div class="box-content">
<pre class="r"><code>genes %&gt;%
  mutate(length=end_position-start_position) %&gt;% 
  group_by(gene_biotype) %&gt;% 
  summarize(mean_length = floor(mean(length)),
            n_gene = n_distinct(ensembl_gene_id))</code></pre>
<pre><code>## # A tibble: 19 × 3
##                          gene_biotype mean_length n_gene
##                                 &lt;chr&gt;       &lt;dbl&gt;  &lt;int&gt;
## 1            3prime_overlapping_ncRNA        3643      1
## 2                           antisense       16677     81
## 3                           IG_V_gene         435      1
## 4                             lincRNA       27550    192
## 5                               miRNA          86     29
## 6                            misc_RNA         145     24
## 7                processed_pseudogene         818    141
## 8                processed_transcript       83106      7
## 9                      protein_coding       82332    233
## 10                         pseudogene         103      1
## 11                               rRNA         132      9
## 12                     sense_intronic        6817     17
## 13                  sense_overlapping       27095      8
## 14                             snoRNA         118     15
## 15                              snRNA         116     21
## 16                                TEC        1938     16
## 17   transcribed_processed_pseudogene       49814      3
## 18 transcribed_unprocessed_pseudogene       65976      6
## 19             unprocessed_pseudogene        6128     32</code></pre>
</div>
</div>
<ol start="8" class="example" style="list-style-type: decimal">
<li>What is the most frequent single word in the phenotype description on chromosome 21? Split the column using <code>separate</code>, <code>gather</code> the columns and <code>count</code> in a single <code>dplyr</code> statement.</li>
</ol>
<div class="box solution">
<h2>
Solution
</h2>
<div class="box-content">
<pre class="r"><code>gene_by_exon %&gt;% 
  dplyr::select(phenotype_description) %&gt;% 
  distinct() %&gt;% 
  filter(!is.na(phenotype_description)) %&gt;% 
  separate(phenotype_description , LETTERS[1:15], extra=&quot;drop&quot;, fill=&quot;right&quot;) %&gt;% 
  gather(everything(), key=&quot;position&quot;, value=&quot;word&quot;) %&gt;% 
  mutate(word=tolower(word)) %&gt;% 
  count(word) %&gt;% 
  arrange(desc(n))</code></pre>
<pre><code>## # A tibble: 207 × 2
##          word     n
##         &lt;chr&gt; &lt;int&gt;
## 1        &lt;NA&gt;  1014
## 2    syndrome    22
## 3        type    16
## 4  deficiency    13
## 5          to    12
## 6   autosomal    11
## 7       early     9
## 8    cataract     8
## 9           1     7
## 10    disease     7
## # ... with 197 more rows</code></pre>
</div>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
